pkg/preference/implem.go

```diff
@@ -0,0 +1,564 @@
+
+//TelemetryConsentDescription adds a description for TelemetryConsentSetting
+var ConsentTelemetryDescription = fmt.Sprintf("If true odo will collect telemetry for the user's odo usage (Default: %t)\n\t\t    For more information: https://developers.redhat.com/article/tool-data-collection", DefaultConsentTelemetrySetting)
+
+// This value can be provided to set a seperate directory for users 'homedir' resolution
+// note for mocking purpose ONLY
+var customHomeDir = os.Getenv("CUSTOM_HOMEDIR")
+
+var (
+	// records information on supported parameters
+	supportedParameterDescriptions = map[string]string{
+		UpdateNotificationSetting: UpdateNotificationSettingDescription,
+		NamePrefixSetting:         NamePrefixSettingDescription,
+		TimeoutSetting:            TimeoutSettingDescription,
+		BuildTimeoutSetting:       BuildTimeoutSettingDescription,
+		PushTimeoutSetting:        PushTimeoutSettingDescription,
+		RegistryCacheTimeSetting:  RegistryCacheTimeDescription,
+		EphemeralSetting:          EphemeralDescription,
+		ConsentTelemetrySetting:   ConsentTelemetryDescription,
+	}
+
+	// set-like map to quickly check if a parameter is supported
+	lowerCaseParameters = util.GetLowerCaseParameters(GetSupportedParameters())
+)
+
+// preferenceInfo wraps the preference and provides helpers to
+// serialize it.
+type preferenceInfo struct {
+	Filename   string `yaml:"FileName,omitempty"`
+	Preference `yaml:",omitempty"`
+}
+
+// odoSettings holds all odo specific configurations
+type odoSettings struct {
+	// Controls if an update notification is shown or not
+	UpdateNotification *bool `yaml:"UpdateNotification,omitempty"`
+
+	// Holds the prefix part of generated random application name
+	NamePrefix *string `yaml:"NamePrefix,omitempty"`
+
+	// Timeout for OpenShift server connection check
+	Timeout *int `yaml:"Timeout,omitempty"`
+
+	// BuildTimeout for OpenShift build timeout check
+	BuildTimeout *int `yaml:"BuildTimeout,omitempty"`
+
+	// PushTimeout for OpenShift pod timeout check
+	PushTimeout *int `yaml:"PushTimeout,omitempty"`
+
+	// RegistryList for telling odo to connect to all the registries in the registry list
+	RegistryList *[]Registry `yaml:"RegistryList,omitempty"`
+
+	// RegistryCacheTime how long odo should cache information from registry
+	RegistryCacheTime *int `yaml:"RegistryCacheTime,omitempty"`
+
+	// Ephemeral if true creates odo emptyDir to store odo source code
+	Ephemeral *bool `yaml:"Ephemeral,omitempty"`
+
+	// ConsentTelemetry if true collects telemetry for odo
+	ConsentTelemetry *bool `yaml:"ConsentTelemetry,omitempty"`
+}
+
+// Registry includes the registry metadata
+type Registry struct {
+	Name   string `yaml:"Name,omitempty"`
+	URL    string `yaml:"URL,omitempty"`
+	Secure bool
+}
+
+// Preference stores all the preferences related to odo
+type Preference struct {
+	metav1.TypeMeta `yaml:",inline"`
+
+	// Odo settings holds the odo specific global settings
+	OdoSettings odoSettings `yaml:"OdoSettings,omitempty"`
```

Comment:
Why do we need this struct? Can we not simply implement the `OdoSettings` properties into `Preference`?

Reply:
I think is has been done this way to have this format for the preference file, similar to the Kubernetes resources format (with `OdoSettings` insted of `Spec`):
```
kind: Preference
apiversion: odo.dev/v1alpha1
OdoSettings:
  RegistryList:
  - Name: DefaultDevfileRegistry
    URL: https://registry.devfile.io
    secure: false
  Ephemeral: false
  ConsentTelemetry: false
```

If we inline the fields of OdoSettings into Preferences, all the fields would be at the same level as `kind` and `apiVersion`.